### PR TITLE
Support for recursive references within JSON schemas

### PR DESCRIPTION
**Pull request by tajo**
This PR adds support for recursive references within JSON schemas. The current resolver ends up running forever in cases like this:

``` js
"outerObject": {
  "type": "array",
  "items": {
    "anyOf": [{
      "type": "string",
      "example": "abc"
    }, {
      "$ref": "#/outerObject"
    }]
  }
}
```

In order to detect cycles (recursive/circular references), the algorithm maintains an array with counters Every time when resolver finds a property `$ref`, it makes a lookup in this array (keys are JSON paths). If the value is bigger than 0, it means we are in the cycle and reference is resolved into:

``` js
{
  type: 'string',
  description: '__RECURSIVE',
  example: value + ' in: ' + schemaId
}
```

That being said, I tried some smaller and bigger updates of `resolver.js` to implement this but without success. The biggest problem is fact that everything is mutable and references are continually changing under our hands. (https://github.com/cloudflare/json-schema-docs-generator/blob/master/lib/resolver.js#L79)

So I did a complete rewrite of `resolver.js`. It's not mutating its inputs. I replaced the resolver class by a set of functions that are mostly without side-effects (it's often using _.cloneDeep).
## How it works
- The input (raw) schemas are used as a recipe for recreation of schemas with resolved references.
- The main recursive function `dereferenceSchema` is using map function.
- It doesn't pass down the whole schemas (or subschemas) but only string `path` (JSON pointer).
- It never needs to access parents or children, every level is built separately. It doesn't need to "cheat" and pass a stack as parameter.
- https://www.npmjs.com/package/flattenator library is used to easily remove unnecessary $ref in the output, it transforms this

``` js
"foo": {
  "type": "array",
  "items": {
    "anyOf": [{
      "$ref": {
        "type": "string",
        "example": "abc"
      }
    }]
  }
}
```

into

``` js
"foo": {
  "type": "array",
  "items": {
    "anyOf": [{
      "type": "string",
      "example": "abc"
    }]
  }
}
```

The new `resolver.js` is even shorter than the old one and source code should be more readable since there are less side-effects.

### CODE REVIEW

File: lib/resolver.js [hunk 1]
```diff
@@ -1,237 +1,194 @@
...
+        var normRef = normalizeReference(schemaId, path, schemas);
+        var refNode = pointer.get(schemas[normRef.schemaId], normRef.path);
+        var nodePath = normRef.schemaId + normRef.path;
+
+        if (!_.has(counters, nodePath)) {
+          counters[nodePath] = 0;
+        }
+        // break resolving if $ref in this subschema has been already resolved
+        if (counters[nodePath] > RECURSION_LEVELS) {
+          if (debugLevel > 0) {
```
Comment (toekneestuck) [inline lib/resolver.js:96]: You don't want to use `lib/helpers/debug.js`?

Comment (tajo) [inline lib/resolver.js:96]: This helper seems to be working only in a class context, so I can't. But I will update and unify it in an another PR.

**Comment by tajo** [comment]
**Alright. You can merge it now.** I optimized handling of counters, so the script now finishes under 10s for the whole CF-API. I also made a diff between CF-API schemas using the old resolver and new resolver. They are identical.

In future PRs, we can improve example extractor to try avoid recursive branches (now it just picks the first branch for oneOf any anyOf):

```
{
  type: 'string',
  description: '__RECURSIVE',
  example: value + ' in: ' + schemaId
}
```

Also, we can somehow tweak the example value for recursions (`value + ' in: ' + schemaId`).

Recursion level is now [hardcoded](https://github.com/tajo/json-schema-docs-generator/blob/d5b9759aabc55fe876e0e16980f2c910a3fbbf90/lib/resolver.js#L15), it can be easily turned into a parameter if necessary. I already tested multiple levels of recursions and it seems to be working.

---

Since we are using some non-usual characters in our keys like `.`, `[`, `]` and many others since it can be a regexp, I have forked, improved and published https://github.com/tajo/flattenator (based on https://github.com/brycebaril/node-flatnest) so it can use custom separators (delimiters) than dots and brackets. It is now used by resolver to easily remove `$ref` levels from the result.

**Comment by handrews** [comment]
I believe @tajo solved this in https://github.com/cloudflare/doca, which is our currently supported JSON Schema documentation project.  If anyone was waiting for this, please check out Doca instead.

**Pull request closed without merging**